### PR TITLE
nixos/shells-pkgs: added environment.shellPkgs option

### PR DESCRIPTION
**Pull request by pasqui23**
###### Motivation for this change
Nowadays nixpkgs contains several packages containing snippets to be added to shells initializations, however those snippets are often in disparate places and need to read the package before knowing the specific path to be sourced.

This option allow for a programmatic addition of shell snippets, as simply as `environment.sessionPackages`

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aanderse** [comment]
After a quick read through I'm left wanting an example of when you might use this option. Just out of curiosity.

**Comment by pasqui23** [comment]
@aanderse the reason being that most shell plugins contain the script to be `source`d in several non standard locaation and often one need to look at the specific nix file. My approach instead is as easy as environment.systemPackages.
I've also pushed some changes to packages so that they can use the new framework.

**Comment by aanderse** [comment]
Oh get it now. Thanks for explaining.

**Review by sternenseemann — COMMENTED** [review]
Cool idea! My thoughts are the following:

* The code is a bit hard to read. More comments explaining and less aliases for stuff we know would be nice :)
* I feel like this module is *too* magic. I'd prefer it if `environment.shellPkgs` was namespaced *per Shell*  (i. e. `environment.shellPkgs.fish, …), alternatively `programs."${shell}".shellPkgs` could also be interesting. This would have the following benefits in my eyes:
  * Users generally want to install stuff *for* a specific shell, not specific stuff for all shell which then only supports one shell anyways.
  * If someone wants to install a zsh package, but uses say fish, with `environment.shellPkgs` it may be confusing why it didn't work. If it was namespaced we could add an assertion to make sure that doesn't happen or make evaluation fail.
* The extra output is a neat idea, but also not nice that we duplicate everything effectively. Maybe there is an alternative solution like (ab)using `passthru` and adding a `supportedShells` list or something.
* Needs documentation in the `nixpkgs` manual, so packagers know how to package a shell package to support this.

Just my 2 cents, let me know what you think on those points!

**Comment by pasqui23** [comment]
@sternenseemann Added documentation ,`programs.${sh}.shellPkgs` options and clarified the code as you requested.

File: pkgs/tools/misc/direnv/default.nix [hunk 1]
```diff
@@ -16,15 +15,25 @@ buildGoModule rec {
...
+  outputs = [ "out" ] ++ map (sh: "${sh}_interactiveShellInit");
+  shells = [ "bash" "fish" "zsh" ];
+
   installPhase = ''
-    make install PREFIX=$out
+        make install PREFIX=$out
+        ${lib.concatMapStrings
+    (
+        sh: ''
+          $out/bin/direnv hook ${sh} > $${sh}_interactiveShellInit
```
Comment (sternenseemann) [inline pkgs/tools/misc/direnv/default.nix:33]: `$${foo}` escapes the `${foo}`.

```suggestion
          $out/bin/direnv hook ${sh} > ''$${sh}_interactiveShellInit
```

**Comment by pasqui23** [comment]
Still developing it

**Comment by pasqui23** [comment]
Still waiting on @sternenseemann ,@Ma27 or @mweinelt to review this

File: doc/languages-frameworks/shell.section.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Shell Packages
+Many packages provide some shell script that gives extra functionality if `source`d from the shell.
+Those need to have extra outputs called `${shell}_${phase}` where:
+* shell refers to the specific shell it needs to be included (either bash, fish or zsh).
+* phase refers to when the file will be sourced, this can be one of:
+* * `interactiveShellInit`,if it needs to be included in the interactive shell initialization
```
Comment (SuperSandro2000) [inline doc/languages-frameworks/shell.section.md:6]:
```suggestion
* * `interactiveShellInit` if it needs to be included in the interactive shell initialization
```

File: pkgs/shells/zsh/fzf-zsh/default.nix [hunk 3]
```diff
@@ -23,8 +23,11 @@ stdenv.mkDerivation rec {
 
   dontBuild = true;
 
+  outputs = [ "out" "interactiveShellInit_zsh"];
+
   installPhase = ''
     install -Dm0644 fzf-zsh.plugin.zsh $out/share/zsh/plugins/fzf-zsh/fzf-zsh.plugin.zsh
+    install -Dm0644 fzf-zsh.plugin.zsh $interactiveShellInit_zsh
```
Comment (SuperSandro2000) [inline pkgs/shells/zsh/fzf-zsh/default.nix:30]: I think it would be a better design to only have the interactiveShellInit output and subdirectories for bash, fish and zsh. This would also be easier to extend.

File: pkgs/tools/misc/direnv/default.nix [hunk 4]
```diff
@@ -16,15 +15,25 @@ buildGoModule rec {
...
     lib.optionalString (!stdenv.hostPlatform.isWindows)
-    "${bash}/bin/bash";
+      "${bash}/bin/bash";
 
   # replace the build phase to use the GNUMakefile instead
   buildPhase = ''
     make BASH_PATH=$BASH_PATH
   '';
 
+  outputs = [ "out" ] ++ map (sh: "${sh}_interactiveShellInit");
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/direnv/default.nix:25]: Why do we have once the shells at the end and here at the front? Not very consistent.

File: pkgs/shells/zsh/fzf-zsh/default.nix (hunk 3 above)
Comment (pasqui23) [reply pkgs/shells/zsh/fzf-zsh/default.nix:30]: I've chosen this design for a couple of reason:

1. It allow users of, say, zsh to not pay for fish plugins
2. It allows `programs,zsh,shellPkgs` to contain only packages with zsh plugins
3. In the immediate future I intend to create (either here or on home-manager) a module that allows me to manage zsh plugins to be loaded and managed by zplugin to improve parsing even further.
4. Especially point 3 require IFD with your design

In short I need to filter plugins by shell without IFD and I can't do that with per shell subdirs

File: pkgs/tools/misc/direnv/default.nix (hunk 4 above)
Comment (pasqui23) [reply pkgs/tools/misc/direnv/default.nix:25]: I don't understand

File: pkgs/shells/zsh/fzf-zsh/default.nix (hunk 3 above)
Comment (pasqui23) [reply pkgs/shells/zsh/fzf-zsh/default.nix:30]: As for extensibility, you just have to add to the lists in the imports in https://github.com/NixOS/nixpkgs/blob/a4e97999decd694aba59773ae384d030a97e08e0/nixos/modules/config/shells-pkgs.nix#L71

File: pkgs/tools/misc/direnv/default.nix (hunk 4 above)
Comment (pasqui23) [reply pkgs/tools/misc/direnv/default.nix]: ok now I get it,it was an error

File: nixos/modules/programs/thefuck.nix [hunk 5]
```diff
@@ -24,16 +24,17 @@ in
             The default value is `fuck`, but you can use anything else as well.
           '';
         };
+        recursive = mkEnableOption "recursively applying rules untilthe command runs";
```
Comment (mmlb) [inline nixos/modules/programs/thefuck.nix:27]:
```suggestion
        recursive = mkEnableOption "recursively applying rules until the command runs";
```

**Comment by Daholli** [comment]
@pasqui23 is this something where someone can help you finish this or is this not needed anymore?
